interop-reactiveStreams/shared/src/main/scala/scalaz/zio/interop/reactiveStreams/SinkSubscriber.scala

```diff
@@ -0,0 +1,73 @@
+package scalaz.zio.interop.reactiveStreams
+
+import org.reactivestreams.{ Subscriber, Subscription }
+import scalaz.zio.stream.{ Sink, Stream }
+import scalaz.zio.{ Promise, Queue, Runtime, Task, UIO, ZIO }
+
+class SinkSubscriber[A, B](
+  runtime: Runtime[_],
+  q: Queue[A],
+  p: Promise[Throwable, B]
+) extends Subscriber[A] {
+
+  // all signals in reactive streams are serialized, so we don't need any synchronization
+  private var subscriptionOpt: Option[Subscription] = None
+  private var signalledDemand                       = q.capacity
+
+  def signalDemand: UIO[Unit] =
+    q.size.map { size =>
+      if (size == 0) {
+        val signalDemand = q.capacity - signalledDemand
+        if (signalDemand > 0) subscriptionOpt.foreach(_.request(signalDemand.toLong))
```

Comment:
Let's suspend this side-effect too to make this more resilient to refactoring.

Reply:
done